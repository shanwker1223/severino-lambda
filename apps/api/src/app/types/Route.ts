import { Methods } from './Methods';
import { RouteShorthandOptions } from 'fastify/types/route';
import { FastifyReply, FastifyRequest } from 'fastify';

export default interface Route {
  method: Methods;
  path: string;
  opts: RouteShorthandOptions;

  handler(request: FastifyRequest, reply: FastifyReply);
}
