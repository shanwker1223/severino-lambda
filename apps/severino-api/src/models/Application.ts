import fastify, { FastifyServerOptions } from 'fastify';
import { FastifyInstance } from 'fastify/types/instance';

class Application {
  protected static _app: Application;

  protected fastify: FastifyInstance;

  protected constructor() {
    const fastifyOptions: FastifyServerOptions = { logger: true };
    this.fastify = fastify(fastifyOptions);
    this.fastify.get('/', async () => {
      return 'pong\n';
    });
  }

  public static getInstance(): Application {
    if (typeof Application._app === 'undefined') {
      Application._app = new Application();
    }
    return Application._app;
  }

  public listen(): void {
    this.fastify.listen(
      8080,
      (err, address) => this.serverHandler(err, address)
    );
  }

  protected serverHandler(err: Error, address: string) {
    if (err) {
      console.error(err);
      process.exit(1);
    }
    console.log(`Server listening at ${address}`);
  }
}

export {
  Application
};
