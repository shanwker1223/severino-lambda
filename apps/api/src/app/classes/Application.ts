import fastify, {FastifyReply, FastifyRequest, FastifyServerOptions} from 'fastify';
import {FastifyInstance} from 'fastify/types/instance';
import Module from './Module';
import Route from '../types/Route';
import {Methods} from '../types/Methods';
import {Response as LightMyRequestResponse} from "light-my-request";

export class Application {
  protected static _app: Application;
  private readonly _fastify: FastifyInstance;
  private readonly _routes: Array<Route> = [];

  protected constructor() {
    const fastifyOptions: FastifyServerOptions = {logger: true};
    // noinspection TypeScriptValidateTypes
    this._fastify = fastify(fastifyOptions);
  }

  protected static async createInstance(modules: Array<Module>): Promise<Application> {
    const app = new Application();
    await Promise.all(
      modules.map(async (module: Module): Promise<void> => app.registerModule(module))
    );
    return app;
  }

  protected serverHandler(err: Error, address: string): void {
    if (err) {
      console.error(err);
      throw err;
    }
    console.log(`Server listening at ${address}`);
  }

  public static async getInstance(modules: Array<Module> = []): Promise<Application> {
    if (typeof Application._app === 'undefined') {
      Application._app = await Application.createInstance(modules);
    }
    return Application._app;
  }

  public listen(): void {
    try {
      return this._fastify.listen(
        8080,
        (err: Error | null, address: string): void => this.serverHandler(err, address)
      );
    } catch (err) {
      this._fastify.log.error(err);
      throw err;
    }
  }

  public registerModule(module: Module): void {
    return module.register(this);
  }

  public registerRoute(module: Module, route: Route): void {
    this._routes.push(route);
    switch (route.method) {
      case Methods.get:
        this._fastify.get(
          route.path,
          route.opts,
          (request: FastifyRequest, reply: FastifyReply) => route.handler(request, reply)
        );
    }
  }

  public async inject(method: Methods, route: string): Promise<LightMyRequestResponse> {
    return await this._fastify.inject({
      method: method,
      url: route
    });
  }

  public static destroy(): boolean {
    return delete this._app;
  }

  get fastify(): FastifyInstance {
    return this._fastify;
  }

  get routes(): Array<Route> {
    return this._routes;
  }
}
