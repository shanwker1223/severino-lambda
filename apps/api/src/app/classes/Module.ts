import { Application } from './Application';
import Route from '../types/Route';

export default class Module {
  private readonly _routes : Array<Route>;
  public baseRoute: string;

  constructor(baseRoute: string, routes: Array<Route>) {
    this.baseRoute = baseRoute;
    this._routes = routes;
  }

  get routes(): Array<Route> {
    return this._routes;
  }

  public register(app : Application) : void {
    this._routes.forEach((route : Route) => {
      route.path = this.baseRoute + route.path;
      app.registerRoute(this, route);
    });
  };
}
