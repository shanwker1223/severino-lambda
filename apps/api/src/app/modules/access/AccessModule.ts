import Module from '../../classes/Module';
import NewRoute from './routes/NewRoute';

export default new Module('/access', [NewRoute]);
