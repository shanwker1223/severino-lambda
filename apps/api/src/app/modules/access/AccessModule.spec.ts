import AccessModule from "./AccessModule";
import {Application} from "../../classes/Application";
import Route from "../../types/Route";

describe('Access Module Tests', () => {

  beforeEach(() => {
  });
  afterEach(() => {
  });

  it('AccessModule has same base route', () => {
    expect(AccessModule.baseRoute).toMatchSnapshot();
  });

  it('AccessModule has same routes', () => {
    expect(AccessModule.routes).toMatchSnapshot();
  });

  it('AccessModule registers correctly', async () => {
    const app = await Application.getInstance();

    const registrationFunction = jest.spyOn(app, 'registerRoute');
    AccessModule.register(app);

    expect(app.routes).toMatchSnapshot();

    expect(registrationFunction).toHaveBeenCalledTimes(AccessModule.routes.length);
    AccessModule.routes.forEach((route: Route, index: number) => {
      expect(registrationFunction).toHaveBeenNthCalledWith(index+1, AccessModule, route);
    });

    Application.destroy(); // To prevent application leaks, since it's an singleton
  });
});
