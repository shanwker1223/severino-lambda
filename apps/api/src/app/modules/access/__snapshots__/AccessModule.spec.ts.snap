// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Access Module Tests AccessModule has same base route 1`] = `"/access"`;

exports[`Access Module Tests AccessModule has same routes 1`] = `
Array [
  Object {
    "handler": [Function],
    "method": "get",
    "opts": Object {},
    "path": "/new",
  },
]
`;

exports[`Access Module Tests AccessModule registers correctly 1`] = `
Array [
  Object {
    "handler": [Function],
    "method": "get",
    "opts": Object {},
    "path": "/access/new",
  },
]
`;
